dbuf/dbuf/client.py

```diff
@@ -0,0 +1,167 @@
+import random
+import aiohttp
+import collections
+import struct
+
+import hailtop.utils as utils
+from hailtop.config import get_deploy_config
+
+
+class DBufClient:
+    async def __aenter__(self):
+        self.aiosession = aiohttp.ClientSession(raise_for_status=True,
+                                                timeout=aiohttp.ClientTimeout(total=60))
+        return self
+
+    async def __aexit__(self, exc_type, exc, tb):
+        await self.aiosession.close()
+
+    def __init__(self,
+                 leader_name,
+                 id=None,
+                 max_bufsize=10 * 1024 * 1024 - 1,
+                 deploy_config=None,
+                 rng=None):
+        if not deploy_config:
+            deploy_config = get_deploy_config()
+        if rng is None:
+            rng = random.Random()
+        self.rng = rng
+        self.deploy_config = deploy_config
+        self.root_url = deploy_config.base_url(leader_name)
+        self.session_url = None if id is None else f'{self.root_url}/s/{id}'
+        self.aiosession = None
+        self.id = id
+        self.offs = []
+        self.sizes = []
+        self.cursor = 0
+        self.max_bufsize = max_bufsize
+
+    async def create(self):
+        resp = await utils.request_retry_transient_errors(
+            self.aiosession,
+            'POST',
+            f'{self.root_url}/s')
+        self.id = int(await resp.text())
+        self.session_url = f'{self.root_url}/s/{self.id}'
+        return self.id
+
+    async def start_write(self):
+        workers = await self.get_workers()
+        worker = workers[self.rng.randrange(len(workers))]
+        return DBufAppender(worker, self.deploy_config, self.id, self.max_bufsize, self.aiosession)
+
+    async def get(self, key):
+        server = key[0]
+        server_url = self.deploy_config.base_url(server)
+
+        resp = await utils.request_retry_transient_errors(
+            self.aiosession,
+            'POST',
+            f'{server_url}/s/{self.id}/get',
+            json=key)
+        return await resp.read()
+
+    def _decode(self, byte_array):
+        off = 0
+        result = []
+        while off < len(byte_array):
+            n2 = struct.unpack_from("i", byte_array, off)[0]
+            off += 4
+            result.append(byte_array[off:(off + n2)])
+            off += n2
+        return result
+
+    async def getmany(self, keys):
+        servers = collections.defaultdict(list)
+        results = [None for _ in keys]
+        for i, key in enumerate(keys):
+            servers[key[0]].append((key, i))
+
+        def get_from_server(server, keys):
+            async def f():
+                server_url = self.deploy_config.base_url(server)
+                i = 0
+                while i < len(keys):
+                    batch = []
+                    size = 0
+                    while i < len(keys):
+                        assert keys[i][0][3] < self.max_bufsize
```

Comment:
Nit. Personally, I find `keys[i][0][3]` a little hard to read. Like, I know it's the 3rd field of the 0th field of the ith key, but I wish that we used some kind of name or function to access the field that indicated to me what was stored there so I didn't have to go looking around for it.